### PR TITLE
Some more properties for the tokenstore, uncovering brokenness

### PR DESCRIPTION
**Pull request by christian-marie**
Brokenness is fixed with the merge.

### CODE REVIEW

File: lib/Network/OAuth2/Server/API.hs [hunk 1]
```diff
@@ -566,12 +566,28 @@ serverPostToken
...
-        let link = safeLink (Proxy :: Proxy AnchorOAuth2API) (Proxy :: Proxy ListTokens) page1
-        in throwError err302{errHeaders = [(hLocation, B.pack $ show link)]} --Redirect to tokens page
-    else
-        throwError err500{errBody = "Something quite horrible has happened"}
+    maybe_tok <- liftIO $ storeReadToken ref (Right token_id)
+    tok <- case maybe_tok of
+        Nothing -> throwError err403 { errBody = "Invalid token_id" }
+        Just (_, tok) -> return tok
+    case tokenDetailsUserID tok of
+        Nothing -> do
```
Comment (thsutton) [inline lib/Network/OAuth2/Server/API.hs]: We're not letting an admin revoke non-user-y tokens, then?

File: lib/Network/OAuth2/Server/Store/PostgreSQL.hs [hunk 2]
```diff
@@ -103,31 +103,39 @@ instance TokenStore PSQLConnPool where
...
-                    debugM logName $ "Failed to revoke token " <> show token_id <> " for user " <> show user_id
-                    return False
+                    let msg = "Failed to revoke token " <> show token_id 
+                    errorM logName msg
+                    fail msg
                 x -> do
-                    let errMsg = "Consistency error: revoked multiple (" <> show x <> ") tokens " <> show token_id <> " for user " <> show user_id
-                    errorM logName errMsg
-                    fail errMsg
+                    let msg = "Consistency error: revoked multiple (" <> show x <> ") tokens " <> show token_id
```
Comment (thsutton) [inline lib/Network/OAuth2/Server/Store/PostgreSQL.hs]: It's traditional to mention that "the impossible has happened", no?

File: lib/Network/OAuth2/Server/API.hs [hunk 3]
```diff
@@ -566,12 +566,28 @@ serverPostToken
...
 serverPostToken ref user_id _ (DeleteRequest token_id) = do
-    success <- liftIO $ storeRevokeToken ref user_id token_id
-    if success then
-        let link = safeLink (Proxy :: Proxy AnchorOAuth2API) (Proxy :: Proxy ListTokens) page1
-        in throwError err302{errHeaders = [(hLocation, B.pack $ show link)]} --Redirect to tokens page
-    else
-        throwError err500{errBody = "Something quite horrible has happened"}
+    maybe_tok <- liftIO $ storeReadToken ref (Right token_id)
+    tok <- case maybe_tok of
+        Nothing -> throwError err403 { errBody = "Invalid token_id" }
```
Comment (oswynb) [inline lib/Network/OAuth2/Server/API.hs]: This should be a 404?

File: lib/Network/OAuth2/Server/API.hs (hunk 1 above)
Comment (christian-marie) [inline lib/Network/OAuth2/Server/API.hs]: Well we can, but that needs to be extra logic right?

File: lib/Network/OAuth2/Server/API.hs [hunk 4]
```diff
@@ -566,12 +566,28 @@ serverPostToken
...
+        Just (_, tok) -> return tok
+    case tokenDetailsUserID tok of
+        Nothing -> do
+            liftIO . errorM logName $
+                "user_id " <> show user_id <> " tried to revoke token_id " <>
+                show token_id <> ", which did not have a user_id"
+            throwError err403 { errBody = "Invalid token_id" }
```
Comment (oswynb) [inline lib/Network/OAuth2/Server/API.hs]: Also should probably be 404

File: lib/Network/OAuth2/Server/Store/PostgreSQL.hs (hunk 2 above)
Comment (christian-marie) [inline lib/Network/OAuth2/Server/Store/PostgreSQL.hs]: I will follow convention in future, promise.

File: lib/Network/OAuth2/Server/API.hs [hunk 5]
```diff
@@ -566,12 +566,28 @@ serverPostToken
...
-    if success then
-        let link = safeLink (Proxy :: Proxy AnchorOAuth2API) (Proxy :: Proxy ListTokens) page1
-        in throwError err302{errHeaders = [(hLocation, B.pack $ show link)]} --Redirect to tokens page
-    else
-        throwError err500{errBody = "Something quite horrible has happened"}
+    maybe_tok <- liftIO $ storeReadToken ref (Right token_id)
+    tok <- case maybe_tok of
+        Nothing -> throwError err403 { errBody = "Invalid token_id" }
+        Just (_, tok) -> return tok
+    case tokenDetailsUserID tok of
```
Comment (oswynb) [inline lib/Network/OAuth2/Server/API.hs]: We have a helper function belongsToUser :: TokenDetails -> UserID -> Bool

**Comment by thsutton** [comment]
lgtm

File: lib/Network/OAuth2/Server/API.hs (hunk 4 above)
Comment (christian-marie) [inline lib/Network/OAuth2/Server/API.hs]: It should be exactly the same as any other error, so the client doesn't get extra information it shouldn't. Hence they are all the same. A 404 surely not, I would have said 400 maybe.

File: lib/Network/OAuth2/Server/API.hs [hunk 6]
```diff
@@ -566,12 +566,28 @@ serverPostToken
...
+                show token_id <> ", which did not have a user_id"
+            throwError err403 { errBody = "Invalid token_id" }
+        Just user_id' -> do
+            if user_id == user_id'
+                then liftIO $ storeRevokeToken ref token_id
+                else do
+                    liftIO . errorM logName $
+                        "user_id " <> show user_id <> " tried to revoke token_id " <>
+                        show token_id <> ", which had user_id " <> show user_id'
+                    throwError err403 { errBody = "Invalid token_id" }
```
Comment (oswynb) [inline lib/Network/OAuth2/Server/API.hs]: 404

File: lib/Network/OAuth2/Server/API.hs (hunk 3 above)
Comment (christian-marie) [inline lib/Network/OAuth2/Server/API.hs]: Perhaps a 400, you don't want the user to be able to tell between a valid and invalid token_id, though.

File: lib/Network/OAuth2/Server/Store/PostgreSQL.hs [hunk 7]
```diff
@@ -103,31 +103,39 @@ instance TokenStore PSQLConnPool where
...
-            putStrLn "a"
-            tokens <- query conn "SELECT token_id, token_type, token, expires, user_id, client_id, scope FROM tokens WHERE (user_id = ?) AND revoked is NULL ORDER BY created LIMIT ? OFFSET ?" (uid, size, (p - 1) * size)
-            debugM logName $ "Counting number of tokens for " <> show uid
-            putStrLn "b"
-            [Only numTokens] <- query conn "SELECT count(*) FROM tokens WHERE (user_id = ?)" (Only uid)
-            return (map (\((Only tid) :. tok) -> (tid, tok)) tokens, numTokens)
+            (toks, n_toks) <- case maybe_uid of
+                Nothing -> do
+                    debugM logName "Listing all tokens"
+                    toks <- query conn "SELECT token_id, token_type, token, expires, user_id, client_id, scope FROM tokens WHERE revoked is NULL ORDER BY created LIMIT ? OFFSET ?" (size, (p - 1) * size)
```
Comment (oswynb) [inline lib/Network/OAuth2/Server/Store/PostgreSQL.hs]: nitpick: Have the same base for the two queries and <> in the extra stuff for user_id?

File: lib/Network/OAuth2/Server/API.hs (hunk 3 above)
Comment (oswynb) [inline lib/Network/OAuth2/Server/API.hs]: Exactly, so 404 in both cases?

File: lib/Network/OAuth2/Server/API.hs (hunk 5 above)
Comment (christian-marie) [inline lib/Network/OAuth2/Server/API.hs]: Not when this was written, please feel free to update this PR.

**Comment by christian-marie** [comment]
I just merged #105 into this one, as they both log.

File: lib/Network/OAuth2/Server/Store/PostgreSQL.hs (hunk 7 above)
Comment (christian-marie) [inline lib/Network/OAuth2/Server/Store/PostgreSQL.hs]: Addressed in hipchat, <>ing queries is dangerous.

**Comment by christian-marie** [comment]
oswyn said LGTM

**Pull request merged**